igvc_platform/src/motor_controller/main.cpp

```diff
@@ -1,247 +1,286 @@
-        {
-          ROS_ERROR_STREAM("Recieved unknown string while setting PID values " << ret);
-        }
-      }
-      else if (ret.empty())
-      {
-        ROS_ERROR_STREAM("Empty return from motor board while sending PID values.\t" << ret);
-      }
-    }
-    rate.sleep();
-  }
-  ROS_INFO_STREAM("Sucessfully set all PID values");
+  setPID(sock, rate); // Set PID Values on mbed
 
-  // sends down motor commands and recieves multiple responses back
-  while (ros::ok() && port.isOpen())
+  // send motor commands
+  while (ros::ok())
   {
-    std::string msg = "$" + (enabled ? toBoundedString(current_motor_command.left_velocity) : toBoundedString(0.0)) +
-                      "," + (enabled ? toBoundedString(current_motor_command.right_velocity) : toBoundedString(0.0)) +
-                      "\n";
+      /* This is the buffer where we will store the request message. */
+      uint8_t requestbuffer[256];
+      size_t message_length;
+      bool status;
 
-    port.write(msg);
+      /* allocate space for the request message to the server */
+      RequestMessage request = RequestMessage_init_zero;
 
-    std::string ret = port.readln();
-    size_t dollar = ret.find('$');
-    size_t pound = ret.find('#');
-    int count = 0;
-    while ((dollar != std::string::npos || pound != std::string::npos) && count <= messages_to_read)
-    {
-      count++;
-      // size_t end = ret.find('\n');
-      size_t end = ret.size();
-      std::vector<std::string> tokens;
+      /* Create a stream that will write to our buffer. */
+      pb_ostream_t ostream = pb_ostream_from_buffer(requestbuffer, sizeof(requestbuffer));
+
+      /* indicate that speed fields will contain values */
+      request.has_speed_l = true;
+      request.has_speed_r = true;
+
+      /* fill in the message fields */
+      request.speed_l = static_cast<float>(current_motor_command.left_velocity);
+      request.speed_r = static_cast<float>(current_motor_command.right_velocity);
 
-      if (ret.size() <= 0)
+      /* encode the protobuffer */
+      status = pb_encode(&ostream, RequestMessage_fields, &request);
+      message_length = ostream.bytes_written;
+
+      /* check for any errors.. */
+      if (!status)
       {
-        ROS_INFO_STREAM("Invalid number of tokens from motor board, ret: " << ret);
+          ROS_ERROR_STREAM("Encoding failed: " << PB_GET_ERROR(&ostream));
+          ros::shutdown();
       }
-      else if (pound != std::string::npos)
+
+      /* Send the message strapped to a pigeon's leg! */
+      sock.sendMessage(reinterpret_cast<char*>(requestbuffer), message_length);
+
+      /* Read response from the server */
+      size_t n; // n is the response from socket: 0 means connection closed, otherwise n = num bytes read
+      uint8_t buffer[256]; // buffer to read response into
+
+      memset(buffer, 0, sizeof(buffer));
+      /* read from the buffer */
+      n = sock.readMessage(buffer); // blocks until data is read
+
+      if (n == 0)
```

Comment:
This method is really long. Refactor it into smaller methods.

Reply:
The way this loop is structured currently is as follows: 

1. create a `RequestMessage` and send it to the microcontroller using the "/motors" topic
2. receive a `ResponseMessage` and publish its contents to "/battery", "/robot_enabled", and "/encoders"

I don't think this loop would benefit greatly from being refactored. I say this primarily because there would be a trade-off between decreasing the loop complexity and introducing more global variables. For example, if I were to create a `recieveResponse(...)` method, I would have to either a) make three of the above-mentioned publishers global variables or b) pass the publishers by reference which is bad for scalability.

Reply:
On second thought, i'm going to refactor this entire implementation into a class

Reply:
addressed